File: 238-product-of-array-except-self/main.go
func productExceptSelf(nums []int) []int {
    
    n := len(nums)

	ans := make([]int, n)
	for i := range ans {
		ans[i] = 1
	}

    prefix := 1
	for i := 0; i < n; i++ {
		ans[i] *= prefix
		prefix *= nums[i]
	}

	suffix := 1
	for i := n - 1; i >= 0; i-- {
		ans[i] *= suffix
		suffix *= nums[i]
	}

	return ans
}

File: 238-product-of-array-except-self/main.rs
impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let n = nums.len();
        let mut ans = vec![1; n];
        
        let mut prefix = 1;
        for i in 0..n {
            ans[i] *= prefix;
            prefix *= nums[i];
        }
        
        let mut suffix = 1;
        for i in (0..n).rev() {
            ans[i] *= suffix;
            suffix *= nums[i];
        }
        
        ans
    }
}
